lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+#define uk_store_libid_self()	\
+	uk_store_libid(uk_store_self_name())
+
+/* All basic types that exist - use these when interacting with the API */
+enum uk_store_entry_basic_type {
+	none,
+	s8,
+	u8,
+	s16,
+	u16,
+	s32,
+	u32,
+	s64,
+	u64,
+	uptr
+};
+
+/* Getter definitions */
+typedef __s8 (*uk_store_get_s8_func_t)(void);
+typedef __u8 (*uk_store_get_u8_func_t)(void);
+typedef __s16 (*uk_store_get_s16_func_t)(void);
+typedef __u16 (*uk_store_get_u16_func_t)(void);
+typedef __s32 (*uk_store_get_s32_func_t)(void);
+typedef __u32 (*uk_store_get_u32_func_t)(void);
+typedef __s64 (*uk_store_get_s64_func_t)(void);
+typedef __u64 (*uk_store_get_u64_func_t)(void);
+typedef __uptr (*uk_store_get_uptr_func_t)(void);
+
+/* Setter definitions */
+typedef void (*uk_store_set_s8_func_t)(__s8);
+typedef void (*uk_store_set_u8_func_t)(__u8);
+typedef void (*uk_store_set_s16_func_t)(__s16);
+typedef void (*uk_store_set_u16_func_t)(__u16);
+typedef void (*uk_store_set_s32_func_t)(__s32);
+typedef void (*uk_store_set_u32_func_t)(__u32);
+typedef void (*uk_store_set_s64_func_t)(__s64);
+typedef void (*uk_store_set_u64_func_t)(__u64);
+typedef void (*uk_store_set_uptr_func_t)(__uptr);
+
+struct uk_store_entry {
+	/* Function getter pointer */
+	union {
+		uk_store_get_s8_func_t   s8;
+		uk_store_get_u8_func_t   u8;
+		uk_store_get_s16_func_t  s16;
+		uk_store_get_u16_func_t  u16;
+		uk_store_get_s32_func_t  s32;
+		uk_store_get_u32_func_t  u32;
+		uk_store_get_s64_func_t  s64;
+		uk_store_get_u64_func_t  u64;
+		uk_store_get_uptr_func_t uptr;
+	} get;
+
+	/* Function setter pointer */
+	union {
+		uk_store_set_s8_func_t   s8;
+		uk_store_set_u8_func_t   u8;
+		uk_store_set_s16_func_t  s16;
+		uk_store_set_u16_func_t  u16;
+		uk_store_set_s32_func_t  s32;
+		uk_store_set_u32_func_t  u32;
+		uk_store_set_s64_func_t  s64;
+		uk_store_set_u64_func_t  u64;
+		uk_store_set_uptr_func_t uptr;
+	} set;
+
+	/* The entry name */
+	const char *entry_name;
+
+	/* Entry flags */
+	__u16 flags;
+
+	/* Entry getter/setter type */
+	__u16 type;
+} __align8;
+
+struct uk_store_folder {
+	/* List for the dynamic folders only */
+	struct uk_list_head folder_head;
+
+	/* List for the entries in the folder */
+	struct uk_list_head entry_head;
+
+	/* The folder name */
+	const char *folder_name;
+} __align8;
+
+struct uk_store_folder_entry {
+	/* Saved entry pointer */
+	struct uk_store_entry *entry;
+
+	/* The linked list head */
+	struct uk_list_head list_head;
+
+	/* Refcount */
+	__atomic refcount;
+} __align8;
+
+#define UK_STORE_FLAG_STATIC	1
+
+/**
+ * Retuns the folder_entry of an entry
+ *
+ * @param entry_ptr the entry
+ */
+#define	uk_store_get_folder_entry(entry_ptr)	\
+	__containerof(entry_ptr, struct uk_store_folder_entry, entry)
+
+/* Static entry array start+end points */
+extern struct uk_store_entry *uk_store_entries_start;
+extern struct uk_store_entry *uk_store_entries_end;
+
+/* Do not call directly */
+#define __UK_STORE_INITREG_FOLDER(fldr, fldr_string)			\
```

Comment:
I would avoid folders for the static entries - I think it makes it more complicated and I can imagine that we can live without it. It would have also one benefit when dealing with lookups, like:
``` c
/* lookup for a static entry */
struct uk_store_entry *_uk_store_sentry_find(unsigned int libid, const char *entryname);
/* lookup for a dynamic (at run-time allocated) entry */
struct uk_store_entry *_uk_store_dentry_find(unsigned int libid, const char *foldername, const char *entryname);

/* Do the switch as static inline so that the compiler can optimize to the direct call in case the foldername is hard-coded in the caller code */
static inline struct uk_store_entry *
_uk_store_entry_find(unsigned int libid, const char *foldername, const char *entryname)
{
   if (!foldername)
      return _uk_store_sentry_find(libid, entryname);
   return _uk_store_dentry_find(libid, foldername, entryname);
}

/* API interface */
#define uk_store_entry_find(libname, foldername, entryname) \
   _uk_store_entry_find(uk_store_libid(libname), foldername, entryname)
```

For static entry lookups you just set `foldername` to `NULL`:
``` c
static_entry = uk_store_entry_find(LIBUKALLOC, NULL, "free");
```

Dynamic entries would have a folder name
``` c
dynamic_entry = uk_store_entry_find(LIBUKALLOC, "1", "free");
```

Reply:
Sure, if you think it makes things easier.